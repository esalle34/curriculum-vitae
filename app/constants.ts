enum Pages {
  CV = "cv",
  COMMON = "common",
}

export { Pages };
