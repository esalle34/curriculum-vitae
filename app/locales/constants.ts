enum Languages {
  English = "en",
  Français = "fr",
  "Español" = "es",
}

enum Countries {
  France = "FR",
  USA = "US",
  "United Kingdom" = "GB",
  "España" = "ES",
  Luxembourg = "LU",
  Suisse = "CH",
  Belgique = "BE",
}

const languagesArray = Object.entries(Languages);
const countriesArray = Object.entries(Countries);

const locales = countriesArray.flatMap(([_countryKey, country]) =>
  languagesArray.map(
    ([_languageKey, language]): Locale => `${language}-${country}`
  )
);

export type Locale = `${Languages}-${Countries}`;
const defaultLocale: Locale = `${Languages.English}-${Countries.USA}`;

interface DecodeTranslationOptions {
  className?: string;
  htmlTag: string;
}

export interface DecodeTranslation {
  options: DecodeTranslationOptions;
  translation: string;
}

export {
  Languages,
  Countries,
  defaultLocale,
  languagesArray,
  countriesArray,
  locales,
};
