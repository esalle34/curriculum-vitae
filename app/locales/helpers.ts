import { Countries, DecodeTranslation, Languages, Locale } from "./constants";

const getDictionary = (language: Languages, page: string) => ({
  [language]: () =>
    import(`@/app/locales/${language}/${page}.json`).then(
      (module) => module.default
    ),
});

const getLanguageAndCountry = (locale: Locale) => {
  const [language, country] = locale.split("-");

  return { language: language as Languages, country: country as Countries };
};

const decodeTranslation = ({
  options: { htmlTag, className },
  translation,
}: DecodeTranslation) => {
  return translation
    .replace(`<0>`, `<${htmlTag} ${className ? `class="${className}"` : ""}>`)
    .replace(`</0>`, `</${htmlTag}>`);
};

export { getDictionary, getLanguageAndCountry, decodeTranslation };
