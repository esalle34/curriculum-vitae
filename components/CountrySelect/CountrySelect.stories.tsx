import type { Meta, StoryObj } from "@storybook/react";

import CountrySelect from "./CountrySelect";
import { Countries } from "@/app/locales/constants";

const meta: Meta<typeof CountrySelect> = {
  component: CountrySelect,
};

export default meta;

type Story = StoryObj<typeof CountrySelect>;

/*
 *ðŸ‘‡ Render functions are a framework specific feature to allow you control on how the component renders.
 * See https://storybook.js.org/docs/api/csf
 * to learn how to use render functions.
 */
export const CountrySelectFrance: Story = {
  render: () => (
    <CountrySelect country={Countries.France} setNewCountry={() => {}} />
  ),
};
