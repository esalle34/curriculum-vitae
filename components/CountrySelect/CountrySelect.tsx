"use client";
import { Countries, countriesArray } from "@/app/locales/constants";
import { Dispatch, SetStateAction } from "react";
import getUnicodeFlagIcon from "country-flag-icons/unicode";

interface Props {
  country: Countries;
  setNewCountry: Dispatch<SetStateAction<Countries>>;
}

const CountrySelect = ({ country, setNewCountry }: Props) => (
  <select
    className="my-2 p-2 border-2 border-slate-300 rounded"
    defaultValue={country}
    onChange={(e) => setNewCountry(e.target.value as Countries)}
  >
    {countriesArray.map(([countryKey, countryValue]) => {
      return (
        <option key={countryValue} value={countryValue}>
          {countryKey} - {getUnicodeFlagIcon(countryValue)}
        </option>
      );
    })}
  </select>
);

export default CountrySelect;
