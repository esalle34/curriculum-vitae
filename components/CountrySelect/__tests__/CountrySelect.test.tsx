import { Countries } from "@/app/locales/constants";
import CountrySelect from "..";
import { act, fireEvent, render } from "@testing-library/react";
import getUnicodeFlagIcon from "country-flag-icons/unicode";

describe("CountrySelect", () => {
  it("should render FR snapshot correctly", () => {
    const { container } = render(
      <CountrySelect country={Countries.France} setNewCountry={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });
  it("should render EN snapshot correctly", () => {
    const { container } = render(
      <CountrySelect country={Countries.USA} setNewCountry={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });
  it("should render and work as expected", () => {
    const setNewCountry = jest.fn();
    const { getByRole, getByText } = render(
      <CountrySelect country={Countries.France} setNewCountry={setNewCountry} />
    );
    const select = getByRole("combobox");
    expect(select).toBeInTheDocument();
    fireEvent.change(select, { target: { value: "USA" } });
    const optionUSA = getByText(`USA - ${getUnicodeFlagIcon(Countries.USA)}`);
    expect(optionUSA).toBeInTheDocument();
    expect(setNewCountry).toHaveBeenCalled();
  });
});
