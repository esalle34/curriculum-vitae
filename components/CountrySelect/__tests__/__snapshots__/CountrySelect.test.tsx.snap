// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelect should render EN snapshot correctly 1`] = `
<div>
  <select
    class="my-2 p-2 border-2 border-slate-300 rounded"
  >
    <option
      value="FR"
    >
      France
       - 
      ğŸ‡«ğŸ‡·
    </option>
    <option
      selected=""
      value="US"
    >
      USA
       - 
      ğŸ‡ºğŸ‡¸
    </option>
    <option
      value="GB"
    >
      United Kingdom
       - 
      ğŸ‡¬ğŸ‡§
    </option>
    <option
      value="ES"
    >
      EspaÃ±a
       - 
      ğŸ‡ªğŸ‡¸
    </option>
    <option
      value="LU"
    >
      Luxembourg
       - 
      ğŸ‡±ğŸ‡º
    </option>
    <option
      value="CH"
    >
      Suisse
       - 
      ğŸ‡¨ğŸ‡­
    </option>
    <option
      value="BE"
    >
      Belgique
       - 
      ğŸ‡§ğŸ‡ª
    </option>
  </select>
</div>
`;

exports[`CountrySelect should render FR snapshot correctly 1`] = `
<div>
  <select
    class="my-2 p-2 border-2 border-slate-300 rounded"
  >
    <option
      selected=""
      value="FR"
    >
      France
       - 
      ğŸ‡«ğŸ‡·
    </option>
    <option
      value="US"
    >
      USA
       - 
      ğŸ‡ºğŸ‡¸
    </option>
    <option
      value="GB"
    >
      United Kingdom
       - 
      ğŸ‡¬ğŸ‡§
    </option>
    <option
      value="ES"
    >
      EspaÃ±a
       - 
      ğŸ‡ªğŸ‡¸
    </option>
    <option
      value="LU"
    >
      Luxembourg
       - 
      ğŸ‡±ğŸ‡º
    </option>
    <option
      value="CH"
    >
      Suisse
       - 
      ğŸ‡¨ğŸ‡­
    </option>
    <option
      value="BE"
    >
      Belgique
       - 
      ğŸ‡§ğŸ‡ª
    </option>
  </select>
</div>
`;
